File: dotnet/samples/GettingStarted/Workflow/Step01a_Sequential.cs
// Copyright (c) Microsoft. All rights reserved.

using Microsoft.Agents.Workflows;
using Microsoft.Agents.Workflows.Reflection;

namespace Workflow;

/// <summary>
/// This class demonstrates a simple sequential workflow with two executors.
/// Executors and edges are basic building blocks of workflows. In this sample,
/// the workflow consists of an UppercaseExecutor followed by a ReverseTextExecutor.
/// The UpperCaseExecutor converts input text to uppercase, and the ReverseTextExecutor
/// reverses the text. The two executors are connected via an edge. Data flows from
/// one executor to the next through these edges. Together, the resulting workflow
/// processes the input text by first converting it to uppercase and then reversing it.
/// </summary>
public class Step01a_Sequential(ITestOutputHelper output) : WorkflowSample(output)
{
    [Fact]
    public async Task RunAsync()
    {
        // Create the executors
        UppercaseExecutor uppercase = new();
        ReverseTextExecutor reverse = new();

        // Build the workflow
        WorkflowBuilder builder = new(uppercase);
        builder.AddEdge(uppercase, reverse);
        var workflow = builder.Build<string>();

        // Execute the workflow
        Run run = await InProcessExecution.RunAsync(workflow, "Hello, World!");
        foreach (WorkflowEvent evt in run.NewEvents)
        {
            if (evt is ExecutorCompleteEvent executorComplete)
            {
                Console.WriteLine($"{executorComplete.ExecutorId}: {executorComplete.Data}");
            }
        }
    }
    private sealed class UppercaseExecutor() : ReflectingExecutor<UppercaseExecutor>("UppercaseExecutor"), IMessageHandler<string, string>
    {
        public async ValueTask<string> HandleAsync(string message, IWorkflowContext context)
        {
            string result = message.ToUpperInvariant();
            return result;
        }
    }

    private sealed class ReverseTextExecutor() : ReflectingExecutor<ReverseTextExecutor>("ReverseTextExecutor"), IMessageHandler<string, string>
    {
        public async ValueTask<string> HandleAsync(string message, IWorkflowContext context)
        {
            char[] charArray = message.ToCharArray();
            System.Array.Reverse(charArray);
            string result = new(charArray);

            await context.AddEventAsync(new WorkflowCompletedEvent(result)).ConfigureAwait(false);
            return result;
        }
    }
}

File: python/samples/getting_started/workflow/step_01a_sequential.py
# Copyright (c) Microsoft. All rights reserved.

import asyncio

from agent_framework.workflow import (
    WorkflowBuilder,
    WorkflowCompletedEvent,
    WorkflowContext,
    executor,
)

"""
Sequential Workflow (basic)

What it does:
- Defines two methods decorated with the `@executor` decorator.
- The two methods are are run sequentially: uppercase then reverse.
- Emits a WorkflowCompletedEvent with the final result.

Prerequisites:
- No external services required.
"""


# Step 1: Define methods using the executor decorator.
@executor(id="upper_case_executor")
async def to_upper_case(text: str, ctx: WorkflowContext[str]) -> None:
    """Execute the task by converting the input string to uppercase."""
    result = text.upper()

    # Send the result to the next executor in the workflow.
    await ctx.send_message(result)


@executor(id="reverse_text_executor")
async def reverse_text(text: str, ctx: WorkflowContext[str]) -> None:
    """Execute the task by reversing the input string."""
    result = text[::-1]

    # Send the result with a workflow completion event.
    await ctx.add_event(WorkflowCompletedEvent(result))


async def main():
    """Main function to run the workflow."""
    # Step 2: Build the workflow with the defined edges.
    workflow = WorkflowBuilder().add_edge(to_upper_case, reverse_text).set_start_executor(to_upper_case).build()

    # Step 3: Run the workflow with an initial message.
    events = await workflow.run("hello world")
    print(events.get_completed_event())

    """
    Sample Output:

    WorkflowCompletedEvent(data=DLROW OLLEH)
    """


if __name__ == "__main__":
    asyncio.run(main())
